src/interface/control_enum.hpp

```diff
@@ -0,0 +1,296 @@
+    }
+  }
+}
+
+template<class T>
+bool
+ControlEnum<T>::on_mouse_button_up(const SDL_MouseButtonEvent& button)
+{
+  if (button.button == SDL_BUTTON_LEFT) {
+    Vector mouse_pos = VideoSystem::current()->get_viewport().to_logical(button.x, button.y);
+    if (m_rect.contains(mouse_pos)) {
+      m_open_list = !m_open_list;
+      m_has_focus = true;
+      return true;
+    } else if (Rectf(m_rect.get_left(),
+               m_rect.get_top(),
+               m_rect.get_right(),
+               m_rect.get_bottom() + m_rect.get_height() * float(m_options.size())
+              ).contains(mouse_pos) && m_open_list) {
+      return true;
+    } else {
+      return false;
+    }
+  } else {
+    return false;
+  }
+}
+
+template<class T>
+bool
+ControlEnum<T>::on_mouse_button_down(const SDL_MouseButtonEvent& button)
+{
+  Vector mouse_pos = VideoSystem::current()->get_viewport().to_logical(button.x, button.y);
+  if (m_open_list) {
+    if (!Rectf(m_rect.get_left(),
+               m_rect.get_top(),
+               m_rect.get_right(),
+               m_rect.get_bottom() + m_rect.get_height() * float(m_options.size())
+              ).contains(mouse_pos)) {
+      m_has_focus = false;
+      m_open_list = false;
+    } else {
+      int pos = int(floor((mouse_pos.y - m_rect.get_bottom()) / m_rect.get_height()));
+      if (pos != -1) {
+        // This verification shouldn't be needed but I don't trust myself
+        if (pos >= 0 && pos < int(m_options.size())) {
+          // Yes. We need this. Because I can't acced by numerical index.
+          // There's probably a way, but I'm too bored to investigate.
+          for (auto option : m_options) {
+            if (--pos != -1) continue;
+            *m_value = option.first;
+
+            if (m_on_change)
+              (*m_on_change)();
+
+            break;
+          }
+        } else {
+          log_warning << "Clicked on control enum inside dropdown but at invalid position ("
+                      << pos << " for a size of " << m_options.size() << ")" << std::endl;
+        }
+      }
+      return true;
+    }
+  } else {
+    if (!m_rect.contains(mouse_pos)) {
+      m_has_focus = false;
+      m_open_list = false;
+    }
+  }
+  return false;
+}
+
+template<class T>
+bool
+ControlEnum<T>::on_mouse_motion(const SDL_MouseMotionEvent& motion)
+{
+  InterfaceControl::on_mouse_motion(motion);
+
+  m_mouse_pos = VideoSystem::current()->get_viewport().to_logical(motion.x, motion.y);
+  return false;
+}
+
+template<class T>
+bool
+ControlEnum<T>::on_key_up(const SDL_KeyboardEvent& key)
+{
+  if ((key.keysym.sym == SDLK_SPACE
+    || key.keysym.sym == SDLK_RETURN
+    || key.keysym.sym == SDLK_RETURN2) && m_has_focus) {
+    m_open_list = !m_open_list;
+    return true;
+  } else {
+    return false;
+  }
+}
+
+template<class T>
+bool
+ControlEnum<T>::on_key_down(const SDL_KeyboardEvent& key)
+{
+  if (key.keysym.sym == SDLK_DOWN && m_has_focus) {
```

Comment:
Invert if statements.

Reply:
There's an `else if` in the middle; I put the `m_has_focus` check on top, though.